### DISCUSSION TITLE
Conditional loop using survey-multi-choice plugin

### DISCUSSION THREAD

**Discussion by miriamdietz** [jspsych/jsPsych#2646] [Q&A] [answered]
Hi everyone, 
I am currently programming an online experiment using the  jsPsych version 7.1.2. 
I would like to include comprehension checks in a multiple choice format (using the survey-multi-choice plugin) but I am struggling with creating a loop that takes the participant back to the comprehension check if they chose the wrong answer. 

When executing the code, it always loops the question regardless of the answer and feedback provided. 

Could someone help me out on this? Any tip is appreciated.

Thank you in advance!

Here is the code: 

```js
     //compcheck 1: question and options
     var comprehension_question1 = {
     type: jsPsychSurveyMultiChoice,
     questions: [
        {prompt: '<div align=center><img src="jspsych/img/Like.bmp" width="200px"; height: "200px"></img></div>' +
                 '<div align=center><b>The "thumbs up" symbol indicates that:</b></div>',
          options: [
            'A rater likes me',
            'My prediction was correct',
            'A rater dislikes me',
            'My prediction was incorrect'
          ],

          required: true
        },
      ],
    preamble: "<h3>Please answer the following question.</h3>"
    };


    //compcheck1: this function returns feedback based on response given
    var comp1_feedback = {
      type: jsPsychHtmlButtonResponse,
      stimulus: function(){
        var data = jsPsych.data.getLastTrialData().values()[0]; 
        var answer_Q1 = data.response.Q0; 
        if(answer_Q1.includes('A rater likes me')) { 
          return "<p><img src='jspsych/img/Like.bmp' width='200px'; height: '200px'></img></p>" +
                     "<p align='center'><b>CORRECT!</b> Indeed, this symbol means that a rater likes you and is interested in becoming 
                     friends.</p>";
        } else {
          return "<p align='center'><img src='jspsych/img/Like.bmp' width='200px'; height: '200px'></img></p>" +
                     "<p align='center'><b>INCORRECT.</b> This symbol means that a rater likes you. </p>" 
        }
      },
      choices: ['Next'],
      on_finish: function(data){ 
        if(data.stimulus == "<p align='center'><b>CORRECT!</b> Indeed, this symbol means that a rater likes you and is interested in becoming friends.</p>") {
          data.correct = true; 
        } else {
          data.correct = false; 
        }
      }
    };

    //compcheck1: if answer incorrect, compcheck1 will be repeated until correct response inserted
    var comp1_loop = {
      timeline: [comprehension_question1, comp1_feedback], 
     loop_function: function(data){
     if (data.correct == true) {
       return false
     } else {
       return true
     }
    }
    }
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi @miriamdietz, I'm guessing the scoring isn't working because you're using the [`includes` Javascript function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) to check the response, but `includes` is meant to be called on _arrays_. In your case, you've retrieved the MC option that the participant selected from the `responses` object as a _string_:
```js
var answer_Q1 = data.response.Q0; 
```
So try checking the response string like this:
```js
if(answer_Q1 == 'A rater likes me') { 
```

**Reply by becky-gilbert** [reply]
And if that doesn't work, another tip is to use a `console.log` to check the value of `answer_Q1` to make sure that the value and format is what you're expecting:
```js
var answer_Q1 = data.response.Q0; 
console.log('last response: ', answer_Q1);
if(answer_Q1 == 'A rater likes me') { 
```

**Reply by miriamdietz** [reply]
Hi @becky-gilbert, 

thank you so much for your help and quick reply!

I tried both of your suggestions, but unfortunately neither worked. 
I was thinking that I maybe made a mistake with the looping variable `var comp1_loop = {`, as the comprehension check is still being repeated regardless of the answer being correct or incorrect. However, the question should only be repeated after having inserted the incorrect answer and after having received the feedback that the answer was incorrect, so the participant can try again. 
Can you maybe spot my mistake there?

Thank you in advance. I really appreciate your help!

**Reply by becky-gilbert** [reply]
Oops, sorry! I missed that part. The `data` argument that's passed to the loop function is a data collection. It has the data from the trial(s) that were presented in the last loop. 

So you need to get the values from the data collection, then get the data from the 2nd trial (which is the one that has the accuracy scoring). Try this:
```js
  loop_function: function(data){ 
    if (data.values()[1].correct == true) { 
      return false; 
    } else { 
      return true;
    }
  }
```
Does that work?

**Reply by miriamdietz** [reply]
The loop still does not work and the question keeps looping regardless of the response. 
I tried so many things and am running out of ideas. I still think the problem is in the `loop_function` itself, somehow it does not seem to know what the correct response is.

Do you maybe have another suggestion for me?

The code looks like this now:

    ` //compcheck 1: question and options
       var comprehension_question1 = {
         type: jsPsychSurveyMultiChoice,
         questions: [
            {prompt: '<div align=center><img src="jspsych/img/Like.bmp" width="200px"; height: "200px"></img></div>' +
                     '<div align=center><b>The "thumbs up" symbol indicates that:</b></div>',
              options: [
                'A rater likes me',
                'My prediction was correct',
                'A rater dislikes me',
                'My prediction was incorrect'
              ],
              required: true
            },
          ],
        preamble: "<h3>Please answer the following question.</h3>",
        on_finish: function(data) {
          var answer_Q1 = data.response.Q0; 
            if(answer_Q1 == 'A rater likes me') {
              data.correct = true
            } else {
              data.correct = false
            }
         }
      };

    //compcheck1: this function returns feedback based on response given
    var comp1_feedback = {
      type: jsPsychHtmlButtonResponse,
      stimulus: function(){
        var last_resp_correct = jsPsych.data.getLastTrialData().values()[0].correct; 
        if (last_resp_correct) {
        return "<p><img src='jspsych/img/Like.bmp' width='200px'; height: '200px'></img></p>" +
                   "<p align='center'><b>CORRECT!</b> Indeed, this symbol means that a rater likes you and is interested in becoming friends </p>"
      } else {
        return "<p align='center'><img src='jspsych/img/Like.bmp' width='200px'; height: '200px'></img></p>" +
                   "<p align='center'><b>INCORRECT.</b> This symbol means that a rater likes you. </p>" 
      }},
      choices: ['Next']
    };

    //compcheck1: if answer incorrect, compcheck1 will be repeated until correct response inserted
    var comp1_loop = {
      timeline: [comprehension_question1, comp1_feedback], 
      loop_function: function(data){ 
        if (data.values()[1].correct == true) { 
        return false; 
     } else { 
        return true;
     }
     }
    };
  
    timeline.push(comp1_loop)`

**Reply by jodeleeuw** [reply]
Hi @miriamdietz -- is your experiment available online anywhere? That might help us debug.

**Reply by becky-gilbert** [reply]
Hi @miriamdietz, since you last posted your code, it looks like you moved the accuracy scoring from the 2nd trial in your loop timeline (`comp1_feedback`) to the 1st trial (`comprehension_question1`). So now when you look for the value of `correct` in the loop function, you want to look at the data from the 1st trial by indexing `data.values()` with 0 instead of 1:
```js
loop_function: function(data){ 
  if (data.values()[0].correct == true) {   
```
It might also help for you to look at the data that is passed to the loop function with `console.log`. This will help you figure out where the `correct` property is stored, and check that the responses are being scored correctly:
```js
loop_function: function(data){ 
  console.log('data: ', data.values());
  if (data.values()[0].correct == true) {   
```
Please let us know if that helps!

**Reply by miriamdietz** [reply]
Hi @becky-gilbert @jodeleeuw,

The last tip worked! Thank you so much for your help!

Here is the working code, maybe it is helping someone else, too:

```js
       //compcheck 1: question and options
     var comprehension_question1 = {
     type: jsPsychSurveyMultiChoice,
     questions: [
        {prompt: '<div align=center><img src="jspsych/img/Like.bmp" width="200px"; height: "200px"></img></div>' +
                 '<div align=center><b>The "thumbs up" symbol indicates that:</b></div>',
          options: [
            'A rater likes me',
            'My prediction was correct',
            'A rater dislikes me',
            'My prediction was incorrect'
          ],

          required: true
        },
      ],
    preamble: "<h3>Please answer the following question.</h3>",
    
    on_finish: function(data) {
      //var data = jsPsych.data.getLastTrialData().values()[0]; 
      var answer_Q1 = data.response.Q0; 
        if(answer_Q1 == 'A rater likes me') {
          data.correct = true
        } else {
          data.correct = false
        }}
       };
       
    //compcheck1: this function returns feedback based on response given
    var comp1_feedback = {
      type: jsPsychHtmlButtonResponse,
      stimulus: function(){
        var last_resp_correct = jsPsych.data.getLastTrialData().values()[0].correct; 
        if (last_resp_correct) {
        return "<p><img src='jspsych/img/Like.bmp' width='200px'; height: '200px'></img></p>" +
               "<p align='center'><b>CORRECT!</b> Indeed, this symbol means that a rater likes you and is interested in becoming friends.</p>"
      } else {
        return "<p align='center'><img src='jspsych/img/Like.bmp' width='200px'; height: '200px'></img></p>" +
               "<p align='center'><b>INCORRECT.</b> This symbol means that a rater likes you. </p>" 
      }},
      choices: ['Next']
    };

    //compcheck1: if answer incorrect, compcheck1 will be repeated until correct response inserted
    var comp1_loop = {
      timeline: [comprehension_question1, comp1_feedback], 
      loop_function: function(data){ 
      console.log('data: ', data.values());
      if (data.values()[0].correct == true) {   
      return false; 
    } else { 
      return true;
     }}
    };
    timeline.push(comp1_loop);
```

**Reply by becky-gilbert** [reply]
Yay! Thanks for letting us know 😃
